### DISCUSSION TITLE
Is there a way to export events to syslog/SIEM

### DISCUSSION THREAD

**Discussion by meeas** [evilsocket/opensnitch#638] [Q&A]
I see there is an option in the configs to save database to file.  I assume that database is for the events and that it helps events persist through reboots.  However the file picker doesn't seem to let you select a folder to use to save that file (I assume its a folder it wants) and it always says ":memory:" next to the file picker button.  So this raises a few questions:

 - What type of performance hit do you expect there to be when saving events to a file instead of memory (assuming average workstation usage and perhaps 10 days worth of event with hourly purges).
 - Do you have any build in methods to export this data to syslog so I can offload it to a SIEM for more advanced analysis?  To be honest it would probably just be a locally installed instance of the Elastic Stack for me personally, so just a local log file would also work for my needs, but having a syslog exporter would defiantly increase options for users.
 - Probably WAYYYY out of your scope for this project, but a nice SIEM-like dashboard for your events would be cool to have build in.

**Comment by gustavo-iniguez-goya** [comment]
>  However the file picker doesn't seem to let you select a folder to use to save that file (I assume its a folder it wants)

It expects a file. Select the folder where you want to store the db, and then select an existing file or type the name a new one. It should let you save it.

> What type of performance hit do you expect there to be when saving events to a file instead of memory (assuming average workstation usage and perhaps 10 days worth of event with hourly purges).

As far as I can tell there's no much penalty in terms of speed on v1.5.0, keeping 10 days of events in the DB (~100k records IIRC, it'll depend on the use case). Right now I'm testing 30 days, with 1million of connections events (1GB), purging every 15 minutes (200-4000 deletions), and I don't notice any performance problem inserting/updating events. The DB is stored on a SSD connected via USB3, on a 10 years old laptop.

The memory usage though increases from ~200 MB to ~900 MB. I don't know the inner working of sqlite3 + QSqlDatabase, so I can't give an explanation, but I guess that it'll store some data in memory. Even if you limit the events on the GUI to 50.

Also, there's a small delay when launching the GUI (~1-5 seconds), or when switching limits from 50 to infinity and beyond. But it's much better than on v1.4.0 where it took up to 15 minutes to show up...

The main problem I think is filtering events, especially when joining (just) 2 tables. I'm a bit intrigued about this because it looks like a bug on our side to me, but I haven't had time to debug it. 

On the other hand, if the storage where the events are saved cannot write fast enough, saving events to a file is impossible. I've got 3 servers sending events to an old workstation (Pentium V) and the GUI gets stucked writing events.

> Do you have any build in methods to export this data to syslog so I can offload it to a SIEM for more advanced analysis? To be honest it would probably just be a locally installed instance of the Elastic Stack for me personally, so just a local log file would also work for my needs, but having a syslog exporter would defiantly increase options for users.
> Probably WAYYYY out of your scope for this project, but a nice SIEM-like dashboard for your events would be cool to have build in.

There's no method to export data to other logging platforms for now. But it should be really simple to implement. 

Also for now I don't have in my mind to create a monitoring dashboard. If we saved events to PostgreSQL, MySQL, etc, it would be easier to create one in whatever language/framework. Some time ago I wrote a cli tool, that saves events to PostgreSQL and that may also works only as daemon: https://github.com/gustavo-iniguez-goya/opensnitch/tree/cli/server

**Comment by gustavo-iniguez-goya** [comment]
> Do you have any build in methods to export this data to syslog so I can offload it to a SIEM for more advanced analysis? To be honest it would probably just be a locally installed instance of the Elastic Stack for me personally, so just a local log file would also work for my needs, but having a syslog exporter would defiantly increase options for users.

I think we could integrate logging to syslog easily. The official package supports logging via network, which is a plus.
That way you could filebeat, logstash, graylog, etc to collect and visualize the data.

We could maintain the current format with a new LEVEL or TYPE (like SYSLOG). Then at certain points of the code, write to syslog the relevant events to collect if it's enabled (basically the data displayed on the Events tab)

**Reply by gustavo-iniguez-goya** [reply]
Or to acomodate to the existing parsers for iptables,ufw, etc:

`opensnitch[4044597]:  [SRC="192.168.1.105" DST="18.32.106.11" SPT="123" DPT="123" PROTO="udp" PID="3996570" UID="0" PATH="/usr/sbin/ntpd" CMDLINE="[/usr/sbin/ntpd -p /run/ntpd.pid -c /etc/ntpsec/ntp.conf -g -N -u ntpsec:ntpsec]" CWD="/" ARG1="allow"  ARG2="000-allow-system-cmds"]`

**Reply by gustavo-iniguez-goya** [reply]
collecting data from syslog, promtail and loki, and visualizing it on grafana:

rejects: `{application="opensnitch"} | logfmt | ARG1 =~ "reject"`

![image](https://user-images.githubusercontent.com/2742953/167249501-163fa985-f186-415f-93b6-c86cab2fe0b3.png)

filtering by rule names:
`sum(count_over_time({application="opensnitch"} | logfmt | ARG2 =~ "000-block-domains|000-block-regexp-domains|000-a-pihole-regexp" [5m]) )`

filtering by processes' execution from non-common paths (an unprivileged attacker usually will only have access to world writable directories, /tmp, /var/tmp, /dev/shm, etc):
```
{application="opensnitchd"} | logfmt | PATH !~ "^(/usr/(s|)bin|/usr/lib|/usr/share/grafana).*"
```

More queries:
```
sum(count_over_time({application="opensnitch"} | logfmt | DST !~ "127.*|172.*" [1m]) )
{application="opensnitch"} | logfmt | DST !~ "127.*|10.*|172.*" ARG1 != "allow"
{application="opensnitch"} | logfmt | ARG1 =~ "reject|deny" | line_format "[HOST={{.DSTHOST}} DPT={{.DPT}}]"
{application="opensnitch"} | logfmt | DST !~ "127.*|10.*|172.*|::1" | line_format "{{.PATH}} {{.DST}}"
{application="opensnitch"} | logfmt | DST !~ "127.*|10.*|172.*"
```

![image](https://user-images.githubusercontent.com/2742953/167288109-4f791761-e826-4619-b042-e3a580782e79.png)

**Comment by gustavo-iniguez-goya** [comment]
Added: https://github.com/evilsocket/opensnitch/commit/7b610c017683844de5fceb5cc8819f1a0cae59e4

I'll document how to integrate it with grafana+loki+syslog-ng

**Reply by meeas** [reply]
Awesome!  Thank you!  I will get this up and running this week.

**Comment by gustavo-iniguez-goya** [comment]
I've found a bug/problem where writing to remote syslog via UDP, timeouts for some reason either with syslog-ng or rsyslog, and the connection doesn't recover from that state. 

I've already got a solution working and tested that at least doesn't end into a deadlock and keeps sending events to the server. I'll publish it in a couple of days, but it needs further analysis to determine the real reason behind this behaviour. [edit 26/05/2022: done]

**Comment by pinbuck** [comment]
Outbound firewall button doesn't work, or at least it doesn't stick as intended.
Linux Mint 20.3 (based on ubuntu 20.04)

    "Firewall": "nftables", 
in /etc/opensnitchd/default-config.json was already present by default.
iptables is what was installed by default and I installed nftables via package to see if the buttons would stick after, but no luck.
Where are the entries located for these two buttons if it's possible to manually change them?

https://user-images.githubusercontent.com/96911866/176090477-be4bab99-3203-4c0a-9035-058c9eebfd3e.mp4

Everything else is working just fine other than the new firewall feature. I tested both stable release and am currently using the pre-release.

I set logs to DEBUG after looking through posts and the GUI:
[opensnitchd.log](https://github.com/evilsocket/opensnitch/files/8997703/opensnitchd.log)

I found a whoollle bunch of lines (I was turning it on and off a bunch) that say this, which from my guess might be the issue:
Error parsing firewall configuration /etc/opensnitchd/system-fw.json: json: cannot unmarshal string into Go struct field FwRule.SystemRules.Rule.Position of type uint64

In the meantime I have simply turned the opensnitch firewall frontend off and using iptables ufw gufw.

**Reply by gustavo-iniguez-goya** [reply]
hi @pinbuck, thank you for reporting this problem!

I think this issue was fixed a couple of days ago: https://github.com/evilsocket/opensnitch/commit/1b28fad03f6bfaa684f069d407545b498c161d7e

I'm planning to release new GUI packges with latest fixes, but if you want to test the fix before it, replace your `/usr/lib/python3/dist-packages/opensnitch/dialogs/firewall.py` with https://github.com/evilsocket/opensnitch/blob/1b28fad03f6bfaa684f069d407545b498c161d7e/ui/opensnitch/dialogs/firewall.py

I'll test it again anyway.

**Reply by gustavo-iniguez-goya** [reply]
hey @pinbuck , I think this issue is fixed on latest v1.6.0rc2: https://github.com/evilsocket/opensnitch/releases/tag/v1.6.0-rc.2

**Reply by pinbuck** [reply]
Thanks for the response and all your hard work. I recently learned how to use firejail to block internet for an application but not load any default profiles:
firejail --noprofile --net=none applicationname
and it's proving very helpful. Have yo u used firejail before? I heard some people request firejail stuff.

**Reply by gustavo-iniguez-goya** [reply]
Yes, I use it on a daily basis, mainly to restrict what directories an application can acess to.
